### PR TITLE
Wakeup API support for timeline actions

### PR DESCRIPTION
**Pull request by youtux**

### CODE REVIEW

File: src/js/wakeup/wakeup.js [hunk 1]
```diff
@@ -132,18 +132,21 @@ Wakeup.emitSetResult = function(id, cookie) {
...
   } else {
     e = {
       wakeup: false,
-    };
+      reason: opt.reason,
+      args: opt.args
+    }
   }
 
-  delete this.state.wakeups[id];
```
Comment (youtux) [inline src/js/wakeup/wakeup.js]: I'm not sure why this line was here instead of inside the "event" block.

**Comment by youtux** [comment]
@Meiguro what do you think about this?

**Comment by Meiguro** [comment]
This is good, I've been meaning to merge this with a few changes for awhile now, but supporting 3.0 took quite a bit of time.

The issue has to do with the `.wakeup` bool on the event. This is `true` if it is a wakeup event, and `false if there is no wakeup. `Wakeup.launch` callbacks are always called, even if there is no wakeup event in order to support different logic whether or not there was a wakeup event, so a timeline launch is only distinguished by the existence of `.args`. This can be solved by having a `.timeline` bool, but there are other launch reasons, such as phone, user, and background worker, all of which also don't have bools. With more than one bool, testing for `false` becomes useless. If one would want to have specific logic for when there is no Wakeup event, they would also have to make sure all other type of launch events are false.

Of course, this can be solved by a `.wakeupType`, but I felt that instead I would rather have `Timeline.launch`. I'm planning to merge this in with a small Timeline module that simply handles launch events, similar to the Wakeup module. I'll be able to do so tonight now that pebble.js compiles on 3.0.

**Comment by youtux** [comment]
> ... so a timeline launch is only distinguished by the existence of `.args`

Actually with the proposed implementation you have the `.reason` key in the event object, and that can be one between the `LaunchReasonTypes` (simply-pebble.js:249)

> If one would want to have specific logic for when there is no Wakeup event, they would also have to make sure all other type of launch events are false.

or one could just check that `event.reason !== "wakeup"`.

>  I'm planning to merge this in with a small Timeline module that simply handles launch events, similar to the Wakeup module.

I'm already using this patch in an watchapp that uses the timeline, and my code looks like this:
```javascript
Wakeup.launch(function(e) {
  console.log("Launch reason: " + JSON.stringify(e))
  if (e.reason === "timelineAction") {
    var launchCode = e.args
    dispatchTimelineAction(launchCode)
  }
}
```
Sorry if I am missing a point.

**Comment by youtux** [comment]
I'm giving a look again at the `Wakeup` API, and I think I can see your point. The API is oriented to schedule launches and see if the application has been woken up due to a wakeup event, so attaching the timeline logic is not a good idea.

**Comment by Meiguro** [comment]
Ah, I missed the `.reason` part. I basically spent a week wondering if I wanted `Wakeup` to be the generic launch API for pebble.js. I can already think of one case that would be for that way and is a mess with the multiple modules/callbacks approach: if the user actually does need to determine an action based on multiple launch reasons. I'm hoping this is not a common use case however.

Yes, the general idea is separation of concerns. The fun part with a Timeline module is that there could be helper methods for adding or deleting pins if that isn't straightforward yet. But I want contributions to make it as soon as possible, so it's fine for it to only handle launches at the moment.

**Comment by Meiguro** [comment]
Nice work. It is merged in as `Timeline.launch`. It will pass an event `e` with `e.action` and `e.launchCode`. Launch callbacks are always called at launch, and determining a timeline is by testing against `e.action`. I decided to use launchCode since it would save the confusion of why it's called "args" for those that only use pebble.js and the Timeline API.

I may replace `e.action` with the action object of the pin that launched the app if ever possible, such as if the app itself created the pin. I recommend testing for general existence rather than strictly as a boolean.

**Comment by youtux** [comment]
Great!

**Pull request merged**